src/batchnorm/problem_description.cpp

```diff
@@ -36,6 +36,109 @@ namespace miopen {
+         is_fp16_or_bfp16(bn_problem.GetYDesc().GetType()) &&
+         is_fp32(bn_problem.GetBnScale().GetType()) && is_fp32(bn_problem.GetBnBias().GetType())) ||
+        // case 2 : float type
+        (is_fp32(bn_problem.GetXDesc().GetType()) && is_fp32(bn_problem.GetYDesc().GetType()) &&
+         is_fp32(bn_problem.GetBnScale().GetType()) && is_fp32(bn_problem.GetBnBias().GetType())));
+}
+
+bool IsCKFwdTrainTypeValid(const ProblemDescription& bn_problem)
+{
+    // case 1 : mix type
+    return ((is_fp16_or_bfp16(bn_problem.GetXDesc().GetType()) &&
+             is_fp16_or_bfp16(bn_problem.GetYDesc().GetType()) &&
+             is_fp16_or_bfp16(bn_problem.GetBnScale().GetType()) &&
+             is_fp16_or_bfp16(bn_problem.GetBnBias().GetType()) &&
+             is_fp32(bn_problem.GetBnSMean().GetType()) &&
+             is_fp32(bn_problem.GetBnSVar().GetType())) ||
+            // case 2 : fp32 or fp64
+            (is_fp32_or_fp64(bn_problem.GetXDesc().GetType()) &&
+             is_fp32_or_fp64(bn_problem.GetYDesc().GetType()) &&
+             is_fp32_or_fp64(bn_problem.GetBnScale().GetType()) &&
+             is_fp32_or_fp64(bn_problem.GetBnBias().GetType()) &&
+             is_fp32_or_fp64(bn_problem.GetBnSMean().GetType()) &&
+             is_fp32_or_fp64(bn_problem.GetBnSVar().GetType())));
+}
+
+bool IsOCLBwdTypeValid(const ProblemDescription& bn_problem)
+{
+    return (
+        (is_fp16_or_bfp16(bn_problem.GetXDesc().GetType()) &&
+         is_fp16_or_bfp16(bn_problem.GetDXDesc().GetType()) &&
+         is_fp16_or_bfp16(bn_problem.GetDYDesc().GetType()) &&
+         is_fp32(bn_problem.GetBnScale().GetType()) && is_fp32(bn_problem.GetBnSMean().GetType()) &&
+         is_fp32(bn_problem.GetBnSVar().GetType())) ||
+        // case 1 : fp32
+        (is_fp32(bn_problem.GetXDesc().GetType()) && is_fp32(bn_problem.GetDXDesc().GetType()) &&
+         is_fp32(bn_problem.GetBnScale().GetType()) && is_fp32(bn_problem.GetBnBias().GetType()) &&
+         is_fp32(bn_problem.GetBnSMean().GetType()) && is_fp32(bn_problem.GetBnSVar().GetType())));
+}
+
+bool IsCKBwdTypeValid(const ProblemDescription& bn_problem)
+{
+    return ((is_fp16_or_bfp16(bn_problem.GetXDesc().GetType()) &&
+             bn_problem.GetDXDesc().GetType() == miopenFloat &&
+             is_fp16_or_bfp16(bn_problem.GetBnScale().GetType()) &&
+             bn_problem.GetDYDesc().GetType() == miopenFloat &&
+             bn_problem.GetBnSMean().GetType() == miopenFloat &&
+             bn_problem.GetBnSVar().GetType() == miopenFloat) ||
+            // case 1 : fp32 or fp64
+            (is_fp32_or_fp64(bn_problem.GetXDesc().GetType()) &&
+             is_fp32_or_fp64(bn_problem.GetDXDesc().GetType()) &&
+             is_fp32_or_fp64(bn_problem.GetBnScale().GetType()) &&
+             is_fp32_or_fp64(bn_problem.GetBnBias().GetType()) &&
+             is_fp32_or_fp64(bn_problem.GetBnSMean().GetType()) &&
+             is_fp32_or_fp64(bn_problem.GetBnSVar().GetType())));
+}

```

Comment:
These functions should not be here. ProblemDescription is a universal structure, not tied to a backend, third-party library, etc. Move them to the appropriate files.